### ISSUE TITLE
Kling-Gupta Efficiency metric

### ISSUE THREAD

**Issue by wasf84** [Nixtla/neuralforecast#1270]
### Description

I conduct research in the field of water resources. In this research area, the Kling-Gupta Efficiency (KGE) metric is commonly used to compare the results obtained by the model with observed values. An important aspect of this metric is that it simultaneously combines three components: correlation, variability, and bias. Its formula can be found at: https://en.wikipedia.org/wiki/Kling%E2%80%93Gupta_efficiency.

By incorporating these three aspects at the same time, this metric would theoretically allow the model to better capture the non-linearity present in hydrological data, thereby improving model fitting during the training phase. In contrast, using metrics that rely on the mean, such as MAPE and SMAPE, would tend to shift the predictions toward an average point, which might not be ideal for hydrological applications. A good discussion on this metric can be found at: https://scores.readthedocs.io/en/stable/tutorials/Kling_Gupta_Efficiency.html.

However, I have some questions:

1.Is my implementation correct, considering the NeuralForecast library's workflow?
2. Does it make sense to train models using a more complex metric like this? I understand the reasoning behind using this metric, but I am new to this field and still have many doubts.

Thank you for your attention.

Here is the code. Since PyTorch minimizes the model’s error, I had to adjust the output sign of the metric.

```
from neuralforecast.losses.pytorch import BasePointLoss, _weighted_mean

class KGE(BasePointLoss):
    def __init__(self, horizon_weight=None):
        super(KGE, self).__init__(
            horizon_weight=horizon_weight, outputsize_multiplier=1, output_names=[""]
        )

    def __call__(
        self,
        y: torch.Tensor,
        y_hat: torch.Tensor,
        mask: Union[torch.Tensor, None] = None,
    ):
        mean_y = torch.mean(y, dim=-1, keepdim=True)
        mean_y_hat = torch.mean(y_hat, dim=-1, keepdim=True)
        std_y = torch.std(y, dim=-1, keepdim=True)
        std_y_hat = torch.std(y_hat, dim=-1, keepdim=True)

        cov = torch.mean((y - mean_y) * (y_hat - mean_y_hat), dim=-1, keepdim=True)
        corr = cov / (std_y * std_y_hat + 1e-8) 

        beta = mean_y_hat / mean_y
        gamma = std_y_hat / std_y

        kge = - (1 - torch.sqrt((corr - 1) ** 2 + (beta - 1) ** 2 + (gamma - 1) ** 2))

        weights = self._compute_weights(y=y, mask=mask)
        return _weighted_mean(losses=kge, weights=weights)
```

**Comment by wasf84** [comment]
Forget about it. Training an entire network using this metric is not as good as I thought it could be. There are limitations to using global statistical values. My bad.

**Comment by elephaint** [comment]
Thanks for suggesting and the honest self-review 😄 

Not familiar with the metric, but by the looks of your code, I'd suggest to make sure:
-  `mean_y` and `std_y` cannot be nan (so use torch.nanmean and something similar for std)
- Make sure beta cannot be singular, by adding an epsilon to the denominator.
- Is the sqrt in kge really necessary for the result? 

Appreciate sharing the idea / metric!

**Comment by wasf84** [comment]
Hello @elephaint . Thank you in advance for your attention and for your insights.

"Is the sqrt in kge really necessary for the result?"

Yes, the square root is part of the original formulation of this metric. It is indeed rarely used by researchers in the field of machine learning in general, but it is more commonly employed by hydrology researchers. This is because, in this field, it is very common to analyze the dynamics of river flow and rainfall patterns in a given region. That is why this metric incorporates so much information—correlation, relative variability, and bias.  

Thanks again.
Peace.
<3

**Comment by wasf84** [comment]
> Hi [@wasf84](https://github.com/wasf84) , I am interested to hear more from you regarding this, especially on the limitations of using global statistical values.
> 
> It sounds like you have tested and use the designed metric in Neuralforecast library, would like to hear more about what you have observed.

Hello, @JQGoh 
How's it going?

As I mentioned earlier, the KGE metric combines three aspects: correlation, variability, and bias. For this task, it is necessary to compute characteristics of the training and validation data separately. However, variations in these small data segments can interfere with overall statistics, or conversely, have little to no effect.

Take correlation as an example—one of the components of the KGE metric. Small portions of the training and validation data may exhibit correlation values that differ significantly from those of the full dataset. In my understanding, this leads to instability in gradient descent, consequently affecting the backpropagation process in the neural network.

From what I have observed about the behavior of the KGE metric, it would likely be more effective when incorporated into a broader hybrid metric, similar to what is done in HuberMQLoss. In this metric, for small errors, the function behaves quadratically, while for large errors, it transitions to a linear behavior. This results in a more stable gradient, leading to a more stable training process for the neural network.

When it comes to the loss function in a neural network, as far as I understand, the simpler it is, the more stable the gradient descent process tends to be. Therefore, for my specific problem, the best approach seems to be using a loss function that is robust to outliers and penalizes large deviations in data symmetry. The idea is to train the model with such a function and then evaluate the results using a more complex metric, such as KGE, to analyze the overall performance.

Well, I believe that covers it. Part of what I have written here is based on my understanding from the following resource: https://www.deeplearningbook.org/, particularly the chapters explaining gradient descent, loss functions, and backpropagation. This book is an excellent reference. Any errors or misinterpretations are entirely my own responsibility.

Peace. <3

**Comment by marcopeix** [comment]
Really cool that you tried it out and that you took the time to share your experience here! 

I'll close this for now, but if you would like to contribute a tutorial on how to add a loss metric to neuralforecast, I think it would benefit the whole community, and we might see more contributions on that end. 

Let me know if it sounds interesting, and you always reach out to us on Slack!
